src/librustc_borrowck/borrowck/mir/dataflow/mod.rs

```diff
@@ -0,0 +1,1168 @@
+            // has corresponding bit, "gen'ed" (i.e. set)
+            // here, in dataflow vector
+            zero_to_one(&mut sets.gen_set, *move_index);
+        }
+        let bits_per_block = self.bits_per_block(ctxt);
+        match stmt.kind {
+            repr::StatementKind::Assign(ref lvalue, _) => {
+                // assigning into this `lvalue` kills all
+                // MoveOuts from it, and *also* all MoveOuts
+                // for children and associated fragment sets.
+                let move_path_index = rev_lookup.find(lvalue);
+
+                sets.kill_set.set_bit(move_path_index.idx());
+                on_all_children_bits(path_map,
+                                     move_paths,
+                                     move_path_index,
+                                     |moi| {
+                                         assert!(moi.idx() < bits_per_block);
+                                         sets.kill_set.set_bit(moi.idx());
+                                     });
+            }
+        }
+    }
+
+    fn terminator_effect(&self,
+                         ctxt: &Self::Ctxt,
+                         sets: &mut BlockSets,
+                         bb: repr::BasicBlock,
+                         statements_len: usize)
+    {
+        let &(_tcx, mir, ref move_data) = ctxt;
+        let term = mir.basic_block_data(bb).terminator.as_ref().unwrap();
+        let loc_map = &move_data.loc_map;
+        let loc = Location { block: bb, index: statements_len };
+        debug!("terminator {:?} at loc {:?} moves out of move_indexes {:?}",
+               term, loc, &loc_map[loc]);
+        let bits_per_block = self.bits_per_block(ctxt);
+        for move_index in &loc_map[loc] {
+            assert!(move_index.idx() < bits_per_block);
+            zero_to_one(&mut sets.gen_set, *move_index);
+        }
+    }
+
+    fn propagate_call_return(&self,
+                             ctxt: &Self::Ctxt,
+                             in_out: &mut [usize],
+                             _call_bb: repr::BasicBlock,
+                             _dest_bb: repr::BasicBlock,
+                             dest_lval: &repr::Lvalue) {
+        let move_data = &ctxt.2;
+        let move_path_index = move_data.rev_lookup.find(dest_lval);
+        let bits_per_block = self.bits_per_block(ctxt);
+
+        in_out.clear_bit(move_path_index.idx());
+        on_all_children_bits(&move_data.path_map,
+                             &move_data.move_paths,
+                             move_path_index,
+                             |moi| {
+                                 assert!(moi.idx() < bits_per_block);
+                                 in_out.clear_bit(moi.idx());
+                             });
+    }
+}
+
+impl<'a, 'tcx> MaybeInitializedLvals<'a, 'tcx> {
+    fn update_bits(sets: &mut BlockSets, path: MovePathIndex,
+                   state: super::DropFlagState)
+    {
+        match state {
+            DropFlagState::Dead => {
```

Comment:
To have a struct that contains references to both the gen and kill bits, to allow you to do `bits.set(path.idx(), false)` rather than playing around like that and potentially mixing around `clear`/`set`.